### PR TITLE
Add Ubuntu 22.04 support

### PR DESCRIPTION
**Pull request by skylar2-uw**
#### Pull Request (PR) description
This PR adds support for Ubuntu 22.04, based on the existing support for Ubuntu 20.04. I believe there are already test cases for Ubuntu 20.04 so I would expect they would continue to pass for 22.04.

#### This Pull Request (PR) fixes the following issues
n/a

### CODE REVIEW

**Comment by smortex** [comment]
@skylar2-uw can you please add this OS to `metadata.json` so that the CI runs against it?

**Comment by skylar2-uw** [comment]
> @skylar2-uw can you please add this OS to `metadata.json` so that the CI runs against it?

Oops, sorry, missed that step. Should be good to go now.

**Comment by smortex** [comment]
You can probably ignore the CentOS 7, but other the CI tests seems relevant: either the code or the tests needs to be adjusted.

**Comment by bastelfreak** [comment]
I fixed the CentOS failures in https://github.com/voxpupuli/puppet-snmp/pull/256. Please rebase against our latest master branch.

**Comment by skylar2-uw** [comment]
> I fixed the CentOS failures in #256. Please rebase against our latest master branch.

Thanks, @bastelfreak ! I've rebased and extended the existing 20.04 coverage to 22.04. At least in my local testing there's a failing test case with `File[snmpd.sysconfig]` only containing `SNMPDRUN=yes` but not `SNMPDOPTS='-Lsd -Lf /dev/null -u Debian-snmp -g Debian-snmp -I -smux -p /var/run/snmpd.pid'`, but in production I _do_ see both set in `/etc/default/snmpd` so it's possible there's something off with my rake setup (I'm still a beginner). Let me know if anything else needs tweaking.

File: spec/classes/snmp_init_spec.rb [hunk 1]
```diff
@@ -623,7 +623,7 @@
         }
 
         case facts[:os]['release']['major']
-        when '8', '9', '10', '11', '16.04', '18.04', '20.04'
+        when '8', '9', '10', '16.04', '18.04', '20.04', '22.04'
```
Comment (smortex) [inline spec/classes/snmp_init_spec.rb:626]: Looks like you removed the `11` by mistake :wink:

Comment (skylar2-uw) [reply spec/classes/snmp_init_spec.rb]: Gah, you're right, it's been one of those weeks... Fixed and pushed

**Comment by smortex** [comment]
Nice! Can you please combine your commits in a single one?

From your working directory:

```
git rebase -i origin/master # Interactively rewrite history
```

This will bring your editor listing all commits in your branch.  Keep the first line unchanged, and replace `pick` with `squash` for all subsequent lines.  Save and quit, your editor will pop again and allow you to combine all the commit messages in a single one.

Then push your updated branch:

```
git push -f            # Send the changes (-f is required because we re-wrote history)
```

**Comment by skylar2-uw** [comment]
> Nice! Can you please combine your commits in a single one?
> 
> From your working directory:
> 
> ```
> git rebase -i origin/master # Interactively rewrite history
> ```
> 
> This will bring your editor listing all commits in your branch. Keep the first line unchanged, and replace `pick` with `squash` for all subsequent lines. Save and quit, your editor will pop again and allow you to combine all the commit messages in a single one.
> 
> Then push your updated branch:
> 
> ```
> git push -f            # Send the changes (-f is required because we re-wrote history)
> ```

Certainly, thanks a bunch for walking me through this!

**Comment by smortex** [comment]
Oops, it looks like your `origin/master` point to your repo and is not in sync with this one.

No worries, we'll fix that :-)

Assuming `git remote -v` only list `origin` that points to your fork, let's add a remote named 'upstream' and update it:

```
git remote add upstream https://github.com/voxpupuli/puppet-snmp
git fetch upstream
```

Then we can rebase (same as before, your editor will pop.  Don't forget to remove all boilerplate generated by git and just write e.g. "Add Ubuntu 22.04 support" as commit message) and push:

```
git rebase -i upstream/master
git push -f
```

The "commits" tab should then only show one commit :+1: 

Thanks!

**Comment by skylar2-uw** [comment]
> Oops, it looks like your `origin/master` point to your repo and is not in sync with this one.
> 
> No worries, we'll fix that :-)
> 
> Assuming `git remote -v` only list `origin` that points to your fork, let's add a remote named 'upstream' and update it:
> 
> ```
> git remote add upstream https://github.com/voxpupuli/puppet-snmp
> git fetch upstream
> ```
> 
> Then we can rebase (same as before, your editor will pop. Don't forget to remove all boilerplate generated by git and just write e.g. "Add Ubuntu 22.04 support" as commit message) and push:
> 
> ```
> git rebase -i upstream/master
> git push -f
> ```
> 
> The "commits" tab should then only show one commit +1
> 
> Thanks!

OK, rebased and pushed!

Skylar

**Comment by skylar2-uw** [comment]
I'm finally getting back to this, and trying to make sense of the couple test failures that are in here. It looks like one problem is here in [`manifests/init.pp`](https://github.com/uwgsit/puppet-snmp/blob/v600_ubuntu_2204/manifests/init.pp#L514), with this [corresponding test case](https://github.com/voxpupuli/puppet-snmp/blob/bb0f9a6c9a8e6d2512926852796429b1f2e3deea/spec/classes/snmp_init_spec.rb#L575). I think the problem might be that the Ubuntu 22.04 package has `TRAPDOPTS` already set. Do we want to remove/disable that from the package, or update the test case not to care for Ubuntu 22.04+?

For the second failure, it looks like it's a user mismatch (`snmp` vs `Debian-snmp`) but I'm not sure if the intent for the module is to use its own user (`snmp`) or the OS user (`Debian-snmp`). Let me know and I think I can figure out how to tweak things accordingly.

Thanks!

**Comment by kenyon** [comment]
In general we want to keep things as close to the distribution packages as possible.

**Comment by skylar2-uw** [comment]
> In general we want to keep things as close to the distribution packages as possible.

Thanks for the clarification! I think I've got the Ubuntu 22.04 test cases fixed, though there's now some problems with Debian 11. I don't think I would have done anything that would affect that, but unfortunately don't have a Debian box to test on. Let me know if there's anything I can do to track down the problem, though.

**Comment by kenyon** [comment]
@skylar2-uw maybe this broke Debian 11: https://github.com/voxpupuli/puppet-snmp/pull/268#discussion_r1117811563

**Comment by skylar2-uw** [comment]
Thanks for catching that, should be good to go now!

**Review by kenyon — APPROVED** [review]
Needs a rebase and better commit message, but otherwise seems fine.

**Comment by skylar2-uw** [comment]
> Needs a rebase and better commit message, but otherwise seems fine.

Thanks, @kenyon! I've rebased and set a more descriptive commit message.

**Comment by skylar2-uw** [comment]
> You want to rebase against upstream/master. I just did it and it looks good, but I can't push to your branch. If you change the permissions on this pull request so that maintainers can push to it, I can fix it.

Unfortunately I'm not seeing that option, I think because the fork is in our org account rather than a personal account. Is it sufficient just to re-do the rebase against `upstream/master`?

**Comment by kenyon** [comment]
Yes, that's all I did, `git fetch upstream; git rebase upstream/master`.

**Comment by skylar2-uw** [comment]
> Yes, that's all I did, `git fetch upstream; git rebase upstream/master`.

Thanks! Think I have all that sorted now. Sorry for the trouble, old CVS/SVN hand trying to get used to the `git` world...

**Comment by kenyon** [comment]
There are still unrelated commits. I think it's because your upstream/master is not up to date.

**Comment by skylar2-uw** [comment]
Weird... `git fetch upstream` reports no changes, and `git remote -v|grep ^upsteram` looks OK
```
upstream        https://github.com/voxpupuli/puppet-snmp (fetch)
upstream        https://github.com/voxpupuli/puppet-snmp (push)
```

Is there something obvious I'm missing? Worst case I could re-create the PR against the current master, though if there's a process I can improve for next time, let me know.

**Comment by skylar2-uw** [comment]
Aha, think I figured it out, was doing `git push` rather than `git push -f`. Think it's working now!

**Review by smortex — APPROVED** [review]
LGTM!

**Review by kenyon — APPROVED** [review]

**Pull request merged**